solidity/core/test/validator-manager/outboxValidatorManager.test.ts

```diff
@@ -39,82 +84,379 @@ describe('OutboxValidatorManager', () => {
+      // empty leaf at leafIndex, but that's enough for us.
+      const root = await helperOutbox.branchRoot(leaf, proof, leafIndex);
+      const index = messageCount - 1;
+      const signatures = await signCheckpoint(
+        root,
+        index,
+        [validator0, validator1], // 2/2 signers is a quorum
+      );
       await expect(
-        validatorManager.improperCheckpoint(
+        validatorManager.invalidCheckpoint(
           outbox.address,
           root,
           index,
           signatures,
+          leaf,
+          proof,
+          leafIndex,
         ),
       )
-        .to.emit(validatorManager, 'ImproperCheckpoint')
+        .to.emit(validatorManager, 'InvalidCheckpoint')
         .withArgs(outbox.address, root, index, signatures);
       expect(await outbox.state()).to.equal(types.AbacusState.FAILED);
     });
 
-    it('reverts if there is not a quorum', async () => {
+    it('reverts if an invalidity proof of a non-empty leaf is not signed by a quorum', async () => {
+      const invalid = await getProofForDispatchedMessage(helperOutbox, message);
+      const root = await helperOutbox.root();
+      const index = messageCount - 1;
       const signatures = await signCheckpoint(
         root,
         index,
         [validator0], // 1/2 signers is not a quorum
       );
-
       await expect(
-        validatorManager.improperCheckpoint(
+        validatorManager.invalidCheckpoint(
           outbox.address,
           root,
           index,
           signatures,
+          invalid.leaf,
+          invalid.proof,
+          invalid.index,
         ),
       ).to.be.revertedWith('!quorum');
     });
 
-    it('reverts if the checkpoint is not improper', async () => {
-      const message = `0x${Buffer.alloc(10).toString('hex')}`;
-      // Send two messages to allow checkpointing
-      await outbox.dispatch(
-        INBOX_DOMAIN,
-        utils.addressToBytes32(signer.address),
-        message,
+    it('reverts if the signed root does not match the invalidity proof', async () => {
+      const root = await helperOutbox.root();
+      const invalid = await getProofForDispatchedMessage(helperOutbox, message);
+      const index = messageCount - 1;
+      const signatures = await signCheckpoint(
+        root,
+        index,
+        [validator0, validator1], // 2/2 signers is not a quorum
       );
-      await outbox.dispatch(
-        INBOX_DOMAIN,
-        utils.addressToBytes32(signer.address),
-        message,
+      await expect(
+        validatorManager.invalidCheckpoint(
+          outbox.address,
+          root,
+          index,
+          signatures,
+          invalid.leaf,
+          invalid.proof,
+          invalid.index,
+        ),
+      ).to.be.revertedWith('!root');
+    });
+
+    it('reverts if the proved leaf is valid', async () => {
+      const valid = await getProofForDispatchedMessage(helperOutbox, message);
```

Comment:
How is this message valid? IIUC, shouldn't it be invalid as `outbox` has no messages enqueued and thus this checkpoint is ahead of the outbox?

Reply:
That's the definition of `premature`, not `invalid`. You'll see that the definition of `invalid` does not actually depend on the contents of the Outbox.

Reply:
I see how that is technically correct, but I find the use of the word `valid` in this context still confusing fwiw

Reply:
No longer applicable as we now only have `premature` and `fraudulent` checkpoints